exercises/nucleotide-count/example.go

```diff
@@ -5,35 +5,37 @@ import (
 	"strings"
 )
 
-// Histogram is a mapping from nucleotide to its count in given DNA
-type Histogram map[byte]int
-
 // DNA is a list of nucleotides
 type DNA string
+type Histogram map[byte]int
 
 const validNucleotides = "ACGT"
 
+func isValidNucleotide(nucleotide byte) bool {
+	return strings.Contains(validNucleotides, string(nucleotide))
+}
+
 // Count counts number of occurrences of given nucleotide in given DNA
 func (dna DNA) Count(nucleotide byte) (count int, err error) {
-	if !strings.Contains(validNucleotides, string(nucleotide)) {
+	if !isValidNucleotide(nucleotide) {
 		return 0, errors.New("dna: invalid nucleotide " + string(nucleotide))
 	}
-
 	return strings.Count(string(dna), string(nucleotide)), nil
 }
 
 // Counts generates a histogram of valid nucleotides in given DNA.
 // Returns error if DNA contains invalid nucleotide.
-func (dna DNA) Counts() (Histogram, error) {
+func (dna DNA) Counts() (result Histogram, e error) {
 	var total int
-	h := Histogram{}
+	result = make(Histogram)
+
 	for i := range validNucleotides {
 		nucleotide := validNucleotides[i]
-		h[nucleotide], _ = dna.Count(nucleotide)
-		total += h[nucleotide]
+		result[nucleotide], _ = dna.Count(nucleotide)
```

Comment:
Perhaps replace
> result[nucleotide], _ = dna.Count(nucleotide)

with
> if result[nucleotide], e = dna.Count(nucleotide); e != nil {
>    return nil, e
> }

Then var total is unnecessary as well as the later if test using it; it also preserves the specific error from Count.

Reply:
That would fail the test with invalid nucleotides, like "AGXXACT" in the test suite... Because the algorithm only iterates on valid ones "ACGT"

Reply:
Ah, I see now.